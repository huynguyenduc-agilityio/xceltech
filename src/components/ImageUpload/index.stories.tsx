import { Meta, StoryObj } from '@storybook/react';

// Components
import ImageUpload from '.';

const meta = {
  title: 'Components/ImageUpload',
  component: ImageUpload,
  parameters: {
    layout: 'centered',
  },
  render: (arg) => (
    <div className="flex justify-center w-[1000px] p-9 bg-white">
      <ImageUpload {...arg} />
    </div>
  ),
  tags: ['autodocs'],
} satisfies Meta<typeof ImageUpload>;

export default meta;

type Story = StoryObj<typeof ImageUpload>;

export const Default: Story = {
  args: {
    onImageChange: () => {},
  },
};
