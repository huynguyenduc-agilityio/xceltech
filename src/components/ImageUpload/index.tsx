import { memo, useCallback, useRef, useState } from 'react';
import { useForm } from 'react-hook-form';

// Icons
import { UserIcon } from '@/icons';
import { PlusIcon } from 'lucide-react';

import {
  Avatar,
  Button,
  FormField,
  FormItem,
  FormMessage,
  Input,
} from '../common';

export interface TImageUploadProps {
  imageUrl?: string;
  onImageChange: (image: string) => void;
}

const ImageUpload = ({ imageUrl = '' }: TImageUploadProps) => {
  const [selectedImageUrl] = useState<string | null>(imageUrl);
  const imageInputRef = useRef<HTMLInputElement>(null);
  const {
    control,
    setValue,
    formState: { errors },
  } = useForm<{ image: File }>({ mode: 'onBlur', reValidateMode: 'onChange' });

  const handleOpenImage = () => imageInputRef.current?.click();

  const handleImageChange = useCallback(
    (e: React.ChangeEvent<HTMLInputElement>) => {
      const image = e.target.files?.[0];
      if (image) {
        setValue('image', image, { shouldValidate: true });

        //TODO: Handle Upload image
      }
    },
    [setValue],
  );

  return (
    <div className="flex flex-col items-start gap-4">
      <FormField
        control={control}
        name={'image'}
        render={() => (
          <FormItem>
            <div className="relative flex items-center justify-center w-[196px] h-[196px] rounded-full bg-gray-neutral group overflow-hidden">
              {selectedImageUrl ? (
                <Avatar
                  size={196}
                  src={selectedImageUrl}
                  alt="Uploaded image"
                  className="cursor-pointer"
                  onClick={handleOpenImage}
                />
              ) : (
                <Button
                  className="w-full h-full"
                  variant="ghost"
                  size="icon"
                  onClick={handleOpenImage}
                >
                  <UserIcon
                    width={80}
                    height={80}
                    className="text-white transition-all group-hover:opacity-50"
                  />
                </Button>
              )}
              {/* Overlay */}
              <div className="absolute inset-0 cursor-pointer pointer-events-none bg-black-default/20 opacity-0 group-hover:opacity-100 transition-opacity flex items-center justify-center">
                <PlusIcon width={40} height={40} className="text-white" />
              </div>

              <Input
                type="file"
                ref={imageInputRef}
                className="hidden"
                onChange={handleImageChange}
                accept="image/*"
              />

              {errors?.image && (
                <FormMessage>
                  <div className="flex flex-col gap-4 text-red-500 text-sm">
                    <p>{errors.image.message}</p>
                    <p>Please select again!!!</p>
                  </div>
                </FormMessage>
              )}
            </div>
          </FormItem>
        )}
      />
    </div>
  );
};

export default memo(ImageUpload);
