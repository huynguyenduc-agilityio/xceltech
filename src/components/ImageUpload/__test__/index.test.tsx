import { render } from '@testing-library/react';

import ImageUpload from '..';

describe('ImageUpload component', () => {
  const mockOnFileChange = jest.fn();

  it('should match snapshot', () => {
    const { container } = render(
      <ImageUpload onImageChange={mockOnFileChange} />,
    );

    expect(container).toMatchSnapshot();
  });
});
