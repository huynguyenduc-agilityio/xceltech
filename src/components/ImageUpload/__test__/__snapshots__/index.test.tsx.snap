// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageUpload component should match snapshot 1`] = `
<div>
  <div
    class="flex flex-col items-start gap-4"
  >
    <div
      class="space-y-2 w-full"
    >
      <div
        class="relative flex items-center justify-center w-[196px] h-[196px] rounded-full bg-gray-neutral group overflow-hidden"
      >
        <button
          class="font-productsans inline-flex items-center justify-center whitespace-nowrap text-lg font-bold ring-offset-background transition-all duration-300 ease-in-out outline-none focus-visible:outline-none disabled:opacity-50 disabled:cursor-not-allowed hover:bg-accent hover:text-accent-foreground rounded-full w-full h-full"
        >
          <svg
            class="text-white transition-all group-hover:opacity-50"
            fill="none"
            height="80"
            viewBox="0 0 28 28"
            width="80"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g
              clip-path="url(#a)"
            >
              <path
                d="M14 15.75a7.877 7.877 0 0 0 7.875-7.875A7.877 7.877 0 0 0 14 0a7.877 7.877 0 0 0-7.875 7.875A7.877 7.877 0 0 0 14 15.75Zm7 1.75h-3.013a9.53 9.53 0 0 1-3.987.875 9.549 9.549 0 0 1-3.987-.875H7a7 7 0 0 0-7 7v.875A2.626 2.626 0 0 0 2.625 28h22.75A2.626 2.626 0 0 0 28 25.375V24.5a7 7 0 0 0-7-7Z"
                fill="currentColor"
              />
            </g>
            <defs>
              <clippath
                id="a"
              >
                <path
                  d="M0 0h28v28H0z"
                  fill="currentColor"
                />
              </clippath>
            </defs>
          </svg>
        </button>
        <div
          class="absolute inset-0 cursor-pointer pointer-events-none bg-black-default/20 opacity-0 group-hover:opacity-100 transition-opacity flex items-center justify-center"
        >
          <svg
            class="lucide lucide-plus text-white"
            fill="none"
            height="40"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="40"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M5 12h14"
            />
            <path
              d="M12 5v14"
            />
          </svg>
        </div>
        <input
          accept="image/*"
          autocomplete="off"
          class="font-productsans w-full px-5 py-4 text-lg text-black-smoky placeholder:text-gray-soft outline-none disabled:cursor-not-allowed disabled:opacity-50 bg-blue-light h-[54px] hidden"
          data-testid="input-element"
          type="file"
        />
      </div>
    </div>
  </div>
</div>
`;
